### PR TITLE
Refactor Genotype and GenotypeAnnotation

### PR DESCRIPTION
**Pull request by heuermh**
Starting point for discussion around `Genotype` and `GenotypeAnnotation`.  This is a tear-down-and-rebuild, where `VariantCallingAnnotations` has been removed and the bare minimum put back in.

### CODE REVIEW

File: src/main/resources/avro/bdg.avdl [hunk 1]
```diff
@@ -988,6 +703,97 @@ record VariantAnnotation {
...
+  /**
+   Genotype for this annotation.
+   */
+  union { null, Genotype } genotype = null;
+
+  /**
+   Additional genotype attributes that do not fit into the standard fields above.
+   The values are stored as strings, even for flag, integer, and float types.
+   */
+  map<string> attributes = {};
```
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: We don't require the attributes here; if no attributes are specified, we will default to an empty map. If we wanted to require the user to set attributes, this line would read:

```
map<string> attributes;
```

Our policy is to always allow fields to be nullable. For pure fields, we do this through the nullable syntax. For maps/arrays, we do this by setting the default to an empty map/array.

Comment (akmorrow13) [inline src/main/resources/avro/bdg.avdl]: ah I see. great!

File: src/main/resources/avro/bdg.avdl [hunk 2]
```diff
@@ -988,6 +703,97 @@ record VariantAnnotation {
...
+   the allele values is significant. VCF genotype field reserved key "GT" mapped to Allele enum
+   values.
+   */
+  array<Allele> alleles = [];
+
+  /**
+   True if this genotype is phased.  If true, the order of alleles is significant. VCF genotype field
+   reserved key "GT" allele separators: '|' genotype is phased; '/' genotype is unphased.
+   */
+  union { boolean, null } phased = false;
```
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: `phased` doesn't really make sense without the `phaseSetId`, and I'd argue that it isn't great to have without `phaseSetQuality`/`phasingQuality`.

File: src/main/resources/avro/bdg.avdl [hunk 3]
```diff
@@ -988,6 +703,97 @@ record VariantAnnotation {
...
+  /**
+   Missing allele.
+   */
+  NO_CALL
+}
+
+/**
+ Genotype.
+ */
+record Genotype {
```
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: We lost a lot of pretty important `Genotype` level fields here, like genotype quality, read depth, alt/reference depth, etc. What's the rationale for that?

File: src/main/resources/avro/bdg.avdl [hunk 4]
```diff
@@ -988,6 +703,97 @@ record VariantAnnotation {
...
+
+  /**
+   The variant for this genotype.
+   */
+  union { null, Variant } variant = null;
+
+  /**
```
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: If we're refactoring `Genotype`, it'd be kinda nice to remove the `contigName`/`start`/`end` fields, since those are nested in `variant`. We flattened them a while back due to a performance regression in Parquet when running predicates.

File: src/main/resources/avro/bdg.avdl [hunk 5]
```diff
@@ -988,6 +703,97 @@ record VariantAnnotation {
...
+   True if this genotype is phased.  If true, the order of alleles is significant. VCF genotype field
+   reserved key "GT" allele separators: '|' genotype is phased; '/' genotype is unphased.
+   */
+  union { boolean, null } phased = false;
+}
+
+/**
+ Genotype annotation.
+ */
+record GenotypeAnnotation {
```
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: I prefer having `GenotypeAnnotation` nest in `Genotype` than the other way around. The reason I like nesting `Variant` in `VariantAnnotation` is because `Variant` nests in `Genotype`, and thus doing a join between `Genotype`s and `VariantAnnotation`s makes sense to me. Historically, we had `VariantCallingAnnotations` as a nested record solely for convenience. I.e., for a "finished" genotype dataset like 1KG, you would probably not populate the `VariantCallingAnnotations`, since they're just used to determine whether a genotype call is correct or not, and the "finished" callset should be "correct".

File: src/main/resources/avro/bdg.avdl [hunk 6]
```diff
@@ -988,6 +703,97 @@ record VariantAnnotation {
 }
 
 /**
+ Allele encodings for genotypes.
+ */
+enum Allele {
```
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: +1 on rename and capitalization

**Comment by heuermh** [comment]
> I don't love these changes. I agree with cutting out some of the fields, but the new Genotype record is far too austere for my preferences. Let me take a hack at a PR that is more of a gradual refactor, as a comparison point.

As I stated above, this pull request was just a starting point for discussion.  I plan to gradually add back most if not all of the missing fields.

I started with what I thought were the most uncontroversial and non-complicated fields.

File: src/main/resources/avro/bdg.avdl (hunk 3 above)
Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: See comment below

**Comment by fnothaft** [comment]
> As I stated above, this pull request was just a starting point for discussion. I plan to gradually add back most if not all of the missing fields.

Oops! Got it now. Sorry about the confusion. I put a (lazily named) view of what I think would be good at #109. Let me know your thoughts.

File: src/main/resources/avro/bdg.avdl (hunk 4 above)
Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: Are nested fields still a problem for Parquet?

File: src/main/resources/avro/bdg.avdl (hunk 5 above)
Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: This is part of what I don't understand about the current API.

In conference with @akmorrow13 this afternoon I'm starting to lean more towards removing the *Annotation records altogether in favor of merging the fields onto `Variant` and `Genotype` and nulling them out via projection.

I'd like further discussion around this, if only for my own understanding.

File: src/main/resources/avro/bdg.avdl (hunk 2 above)
Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: Note before this was defined as `phaseSetId != null`.  According to the VCF spec, one can have phased alleles without specifying a phase set ID:

"All phased genotypes that do not contain a PS subfield are assumed to belong to the same phased set."

File: src/main/resources/avro/bdg.avdl (hunk 4 above)
Comment (akmorrow13) [inline src/main/resources/avro/bdg.avdl]: Yes, nested fields are still a big problem, mostly in the case when we are trying to minimize latency (Mango). One of the motivations for putting in `contigName/start/end` fields was because of this latency issue.

File: src/main/resources/avro/bdg.avdl (hunk 2 above)
Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: I don't see `phaseSetQuality` in the VCF spec.  `phasingQuality` is VCF genotype field reserved key "PQ".  Is that what you meant or did you have something else in mind?

File: src/main/resources/avro/bdg.avdl [hunk 7]
```diff
@@ -955,51 +670,199 @@ record VariantAnnotation {
...
+   reserved key GT allele separators: '|' genotype is phased; '/' genotype is unphased.
+   */
+  union { boolean, null } phased = false;
+
+  /**
+   Phase set identifier, if any. Phased genotypes on the same contig with the same phaseSetId
+   are in the same phased set. All phased genotypes that do not have a phaseSetId are assumed to
+   belong to the same phased set. VCF genotype field reserved key PS.
+   */
+  union { null, string } phaseSetId = null;
```
Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: Note that even though the VCF specification types this as Integer, and recommends "a convention of the position of the first variant in the set identifier (although this is not required)", I've seen all sorts of things used in the wild that aren't Integers.  E.g. the GIAB VCF files use [position_ref_alt], `863556_G_A`

ftp://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/release/NA12878_HG001/latest/NA12878_GIAB_highconf_CG-IllFB-IllGATKHC-Ion-Solid-10X_CHROM1-X_v3.3_highconf.vcf.gz

File: src/main/resources/avro/bdg.avdl [hunk 8]
```diff
@@ -955,51 +670,199 @@ record VariantAnnotation {
...
+   */
+  union { null, boolean } filtersApplied = null;
+
+  /**
+   True if all filters for this variant passed. VCF genotype field reserved key FT value PASS.
+   */
+  union { null, boolean } filtersPassed = null;
+
+  /**
+   Zero or more filters that failed for this variant. VCF genotype field reserved key FT.
```
Comment (jpdna) [inline src/main/resources/avro/bdg.avdl]: should "..that failed for this variant" be "..that failed for this genotype call"?
Also applies to above field.

Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: thanks!  fixed it

File: src/main/resources/avro/bdg.avdl (hunk 7 above)
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: I think I'd prefer to stick with Int, as that's what the VCF spec states, and what is IMO correct (i.e., if I were to implement this, I would use an Int "UUID"). That being said, I agree that it is pretty common to see the string phase set ID. I don't know what's the correct thing to do here. If we go with the string, then we emit headers that disagree with the VCF spec, which seems wrong. If we go with Int, then we can't handle VCFs that come in from the wild. If we went with say:

```
union { null, int } phaseSetId = null;
union { null, string } phaseSetStringId = null;
```

Then we'd need to figure out at save whether the user provided string/int/no phase set IDs.

I don't think there's a right/wrong answer, but we see this elsewhere (https://github.com/bigdatagenomics/adam/issues/1213), so we should hash out our philosophy.

Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: Right, we should go with what @heuermh argued over in #1213.  :)

**Comment by heuermh** [comment]
Things got hairy on the last rebase, so I'm not 100% sure this is currently the way I want it

**Comment by fnothaft** [comment]
SGTM. Ping me when you'd like me to make a review pass.

**Comment by heuermh** [comment]
By this point it is rather hard to follow the diff; I'd suggest looking at the whole `bdg.avro` file and the docs produced by running `mvn javadoc:javadoc`.

There are a few remaining open questions in the comments, including several fields previously from `Genotype` and `VariantCallingAnnotation` that aren't genotype field reserved keys in the VCF specifications.  Please make a case to re-add those that are necessary.

File: src/main/resources/avro/bdg.avdl [hunk 9]
```diff
@@ -748,299 +748,304 @@ record Variant {
 }
 
 /**
- An enumeration that describes the allele that corresponds to a genotype.
+ Allele encodings for genotypes.
  */
-enum GenotypeAllele {
+enum Allele {
```
Comment (devin-petersohn) [inline src/main/resources/avro/bdg.avdl]: Is it possible (or even realistic) to get rid of enums? Impala doesn't support importing our parquet formats because of them. I doubt this is what we want, but I wanted to open a discussion about it.

Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: Thoughts for an alternative?

Comment (devin-petersohn) [inline src/main/resources/avro/bdg.avdl]: The suggested fix for Impala was to rewrite using strings instead.

https://issues.apache.org/jira/browse/IMPALA-2525

Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: Well, what is Impala?  It looks like there is a workaround (define table columns yourself and set enum columns as STRING) and a patch linked to that issue.

File: src/main/resources/avro/bdg.avdl [hunk 10]
```diff
@@ -780,39 +780,189 @@ enum Allele {
 record GenotypeAnnotation {
 
   /**
+   Per-sample total alternate read depth. VCF genotype reserved key AD, Number=R, split for
+   multi-allelic sites into single integer values for the reference allele
+   (referenceReadDepth) and the alternate allele (alternateReadDepth, this field).
+   */
+  union { null, int } alternateReadDepth = null;
```
Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: I will need some help thinking about what these Number=R values on `Genotype`s mean and how they should be split.  I wrote the doc here to match that in `Variant` but am not sure it makes sense.

File: src/main/resources/avro/bdg.avdl [hunk 11]
```diff
@@ -780,39 +780,189 @@ enum Allele {
...
+  /**
+   Genotype likelihoods comprised of floating point log10-scaled likelihoods for all possible
+   genotypes given [the set of alleles defined in the REF and ALT fields]. VCF genotype field
+   reserved key GL, Number=G.
+   */
+  array<float> likelihoods = [];
+
+  // GLE Number=1 Type=String
+  // Genotype likelihoods of heterogeneous ploidy, used in presence of uncertain copy number.
+  // Reserved genotype field key in VCF v4.2, dropped in VCF v4.3
```
Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: Should we keep this in even though it was dropped in the draft VCF v4.3 spec?

File: src/main/resources/avro/bdg.avdl [hunk 12]
```diff
@@ -780,39 +780,189 @@ enum Allele {
...
 
+  // SB is not a reserved key in VCF v4.2 or v4.3
   /**
    Per-sample component statistics which comprise the Fisher's Exact Test to detect strand bias.
    If set, the array should contain the number of reads that support the reference allele on the
    forward strand, the reference allele on the reverse strand, the alternate allele on the forward
    strand, and the alternate allele on the reverse strand, in that order. VCF genotype field key SB.
    */
-  array<int> strandBias = [];
+  //array<int> strandBias = [];  Genotype.strandBiasComponents previously
```
Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: Please make a case to re-add any of these commented out fields.

File: src/main/resources/avro/bdg.avdl (hunk 11 above)
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: I would drop.

File: src/main/resources/avro/bdg.avdl (hunk 12 above)
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: I'll take a look. I am -1 on dropping `strandBias`, and I prefer the old name (`strandBiasComponents`). You can't compute the strand bias components without going back to the reads, and the strand bias info is pretty important for variant filtration.

Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: Thanks!  There's no hesitation from me to re-add any of these if they are worth keeping around.  I'm +1 to `strandBiasComponents`.

File: src/main/resources/avro/bdg.avdl [hunk 13]
```diff
@@ -748,299 +748,304 @@ record Variant {
...
    */
-  union { null, boolean } downsampled = null;
+  union { null, float } expectedAlleleDosage = null; // or expectedAlleleCount? see Variant.alleleCount
 
+  // GL Number=G Type=Float
   /**
-   The Wilcoxon rank-sum test statistic of the base quality scores. The base quality
-   scores are separated by whether or not the base supports the reference or the
-   alternate allele.
+   Genotype likelihoods comprised of floating point log10-scaled likelihoods for all possible
```
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: Are we log or log10?

File: src/main/resources/avro/bdg.avdl [hunk 14]
```diff
@@ -748,299 +748,304 @@ record Variant {
...
+  // EC Number=A Type=Integer
+  // Comma separated list of expected alternate allele counts for each alternate allele in the same
+  // order as listed in the ALT field. Typically used in association analyses.
   /**
-   True if the reads covering this site were randomly downsampled to reduce coverage.
+   The expected dosage of the alternate allele in this sample. VCF genotype reserved key EC,
+   Number=A, split for multi-allelic sites into a single float value.
    */
-  union { null, boolean } downsampled = null;
+  union { null, float } expectedAlleleDosage = null; // or expectedAlleleCount? see Variant.alleleCount
```
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: EC is a field I would drop.

File: src/main/resources/avro/bdg.avdl [hunk 15]
```diff
@@ -748,299 +748,304 @@ record Variant {
...
-   -10 log((a + b)! * (c + d)! * (a + c)! * (b + d)! / (a! b! c! d! n!)
-
-   Where n = a + b + c + d.
-   */
-  union { null, float } fisherStrandBiasPValue = null;
+  // GLE Number=1 Type=String
+  // Genotype likelihoods of heterogeneous ploidy, used in presence of uncertain copy number.
+  // Reserved genotype field key in VCF v4.2, dropped in VCF v4.3
 
+  // GP Number=G Type=Float
```
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: +1 to omitting.

File: src/main/resources/avro/bdg.avdl [hunk 16]
```diff
@@ -748,299 +748,304 @@ record Variant {
...
   /**
-   The root mean square of the mapping qualities of reads covering this site.
+   Genotype posterior probabilities in the range 0.0 to 1.0 using the same ordering as the
```
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: IMO all probabilities/likelihoods should be log scaled.

File: src/main/resources/avro/bdg.avdl [hunk 17]
```diff
@@ -748,299 +748,304 @@ record Variant {
...
    */
-  union { null, float } mqRankSum = null;
+  union { null, int } quality = null;
 
+  // HQ Number=2 Type=Integer
   /**
-   The Wilcoxon rank-sum test statistic of the position of the base in the read at this site.
-   The positions are separated by whether or not the base supports the reference or the
-   alternate allele.
+   Haplotype quality, encoded as two phred qualities. VCF genotype field reserved key HQ, Number=2.
```
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: I am inclined to suggest dropping this field as it is underspecified.

File: src/main/resources/avro/bdg.avdl [hunk 18]
```diff
@@ -748,299 +748,304 @@ record Variant {
...
    */
-  array<float> genotypePriors = [];
+  union { null, int } rmsMappingQuality = null;
 
+  // PL Number=G Type=Integer
   /**
-   The log scaled posterior probabilities of the various genotype states at this site,
-   in this sample. The number of elements in this array should be equal to the ploidy at
-   this site, plus 1.
+   Phred-scaled genotype likelihoods rounded to the closest integer, and otherwise defined precisely
```
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: GL and PL are equivalent, just different scales. We should only have one, and IMO it should be GL.

File: src/main/resources/avro/bdg.avdl [hunk 19]
```diff
@@ -748,299 +748,304 @@ record Variant {
...
+   as the genotype likelihoods field. VCF genotype field reserved key PL, Number=G.
    */
-  array<float> genotypePosteriors = [];
+  array<int> phredScaledLikelihoods = [];
 
+  // PQ Number=1 Type=Integer
   /**
-   The log-odds ratio of being a true vs. false variant under a trained statistical model.
-   This model can be a multivariate Gaussian mixture, support vector machine, etc.
+   Phred-scaled probability that alleles are ordered incorrectly, against all other members in
```
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: It should be "ordered correctly", not "ordered incorrectly", right?

File: src/main/resources/avro/bdg.avdl [hunk 20]
```diff
@@ -748,299 +748,304 @@ record Variant {
...
-   Additional feature info that doesn't fit into the standard fields above.
-   They are all encoded as (string, string) key-value pairs.
+   Per-sample component statistics which comprise the Fisher's Exact Test to detect strand bias.
+   If set, the array should contain the number of reads that support the reference allele on the
+   forward strand, the reference allele on the reverse strand, the alternate allele on the forward
+   strand, and the alternate allele on the reverse strand, in that order. VCF genotype field key SB.
    */
-  map<string> attributes = {};
-}
+  //array<int> strandBias = [];  Genotype.strandBiasComponents previously
```
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: Actually, ignore what I was saying earlier. Since you've added ADF/ADR, I would prefer to use the ADF/ADR fields over using SB.

File: src/main/resources/avro/bdg.avdl [hunk 21]
```diff
@@ -748,299 +748,304 @@ record Variant {
...
+  //
+  // Not present in VCF specification yet present in Genotype
 
-  /**
-   The reference contig that this genotype's variant exists on.
-   */
-  union { null, string } contigName = null;
+  // The minimum number of reads seen at this site across samples when joint calling variants.
+  // Analogous to VCF's MIN_DP.
+  //union { null, int } minReadDepth = null;
```
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: Necessary for gVCF support, don't remove.

File: src/main/resources/avro/bdg.avdl [hunk 22]
```diff
@@ -748,299 +748,304 @@ record Variant {
...
 
-  /**
-   The 0-based start position of this genotype's variant on the reference contig.
-   */
-  union { null, long } start = null;
+  // Log scaled likelihoods that we have n non-reference alleles at this site. The number of
+  // elements in this array should be equal to the ploidy at this site, plus 1.
+  //array<float> nonReferenceLikelihoods = [];
```
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: Necessary for gVCF, don't remove.

Also, log or log10?

File: src/main/resources/avro/bdg.avdl [hunk 23]
```diff
@@ -748,299 +748,304 @@ record Variant {
...
 
-  /**
-   The 0-based, exclusive end position of this genotype's variant on the reference contig.
-   */
-  union { null, long } end = null;
+  // We split multi-allelic VCF lines into multiple single-alternate records. This bit is set
+  // if that happened for this record.
+  //union { boolean, null } splitFromMultiAllelic = false;
```
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: This is removed because it is present in `Variant`, correct? If so, SGTM.

File: src/main/resources/avro/bdg.avdl [hunk 24]
```diff
@@ -748,299 +748,304 @@ record Variant {
...
-  union { null, string }  sampleId = null;
+  //
+  // Not present in VCF specification yet present in VariantCallingAnnotations
 
-  /**
-   A description of this sample.
-   */
-  union { null, string }  sampleDescription = null;
+  // True if the reads covering this site were randomly downsampled to reduce coverage.
+  //union { null, boolean } downsampled = null;
```
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: Necessary, don't remove.

File: src/main/resources/avro/bdg.avdl [hunk 25]
```diff
@@ -748,299 +748,304 @@ record Variant {
...
 
-  /**
-   An array describing the genotype called at this site. The length of this
-   array is equal to the ploidy of the sample at this site. This array may
-   reference OTHER_ALT alleles if this site is multi-allelic in this sample.
-   */
-  array<GenotypeAllele> alleles = [];
+  // The Fisher's exact test score for the strand bias of the reference and alternate
+  // alleles.
+  //union { null, float } fisherStrandBiasPValue = null;
```
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: Necessary for genotype filtering, don't remove.

File: src/main/resources/avro/bdg.avdl [hunk 26]
```diff
@@ -748,299 +748,304 @@ record Variant {
...
+ Genotype.
+ */
+record Genotype {
 
   /**
-   The number of reads that show evidence for the reference at this site.
+   The variant for this genotype. May be left unset or set to null for performance reasons.
    */
-  union { null, int } referenceReadDepth = null;
+  union { null, Variant } variant = null;
```
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: If we're refactoring, we might as well promote the core variant fields here.

File: src/main/resources/avro/bdg.avdl (hunk 14 above)
Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: +1

File: src/main/resources/avro/bdg.avdl (hunk 13 above)
Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: That doc text comes from VCF spec.  The phrase `[the set of alleles defined in the REF and ALT fields]` needs expansion to explain how the values are split for multi-allelic sites.

File: src/main/resources/avro/bdg.avdl (hunk 15 above)
Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: Omit GP, or GLE from lines 847-849 above?  GP is a reserved key in the VCF spec.

File: src/main/resources/avro/bdg.avdl (hunk 13 above)
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: Sure, but my question remains. :-p

File: src/main/resources/avro/bdg.avdl (hunk 15 above)
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: Ah sorry for the confusion, my comment went on the wrong line. I was +1 to omitting GLE, not GP.

File: src/main/resources/avro/bdg.avdl (hunk 16 above)
Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: This doc text comes from VCF spec.  They do have some clarification here
https://github.com/samtools/hts-specs/blob/master/VCFv4.3.tex#L486
but I don't guarantee the spec uses those definitions consistently, nor do I argue that the representation in VCF is correct.  If we deviate we need to document the differences in representation and the conversion between them.

File: src/main/resources/avro/bdg.avdl (hunk 18 above)
Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: So if we read in VCF with PL specified, we convert and write out VCF with GL specified?  Reasonable, needs to be documented.

File: src/main/resources/avro/bdg.avdl (hunk 21 above)
Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: `minimumReadDepth` ok?

File: src/main/resources/avro/bdg.avdl (hunk 16 above)
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: No, I know the text comes from the VCF spec. What I am saying is that _our_ representation for probabilities/likelihood should:

1. Be consistent across all probability/likelihood fields
2. Always be log-scaled (and with a consistent base)

I don't care what the VCF spec says the format for the field is. The VCF spec is terribly inconsistent WRT number formatting. We should not be inconsistent, and should handle what VCF does in the converter.

File: src/main/resources/avro/bdg.avdl (hunk 21 above)
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: OK by me

File: src/main/resources/avro/bdg.avdl (hunk 18 above)
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: Probably would want to have some config setting that choses between write PL, write GL (and possibly write both).

File: src/main/resources/avro/bdg.avdl (hunk 26 above)
Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: What do you mean?  Get rid of the `variant` reference?  Add more fields from `Variant` to `Genotype`?  There will be collisions.

File: src/main/resources/avro/bdg.avdl (hunk 23 above)
Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: I don't see `splitFromMultiAllelic` in `Variant` or `VariantAnnotation`

File: src/main/resources/avro/bdg.avdl (hunk 19 above)
Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: This doc text comes from VCF spec.

File: src/main/resources/avro/bdg.avdl (hunk 23 above)
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: Ah sorry, you're right. What's the rationale for removing this here? If you don't have this field and you have a triallelic site that is `GT=1/1` in a sample, you don't know that there's a second alt at the site from looking at the genotype record.

File: src/main/resources/avro/bdg.avdl (hunk 19 above)
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: Sure, but what's our representation?

File: src/main/resources/avro/bdg.avdl (hunk 26 above)
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: Flatten the core variant fields into genotype (`referenceAllele`, `alternateAllele`, etc), as we did with `contigName`, `start`, `end`.

File: src/main/resources/avro/bdg.avdl (hunk 23 above)
Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: Yeah if we need it, let's add it back.  It being removed was a consequence of how I approached this pull request (drop everything and re-add as necessary).  Should it be added back to `Variant` or `VariantAnnotation` as well?

File: src/main/resources/avro/bdg.avdl (hunk 19 above)
Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: We pass it on as is https://github.com/bigdatagenomics/adam/blob/master/adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantContextConverter.scala#L919

File: src/main/resources/avro/bdg.avdl (hunk 26 above)
Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: There's not much then, the `filter` ones will collide, and all the interesting bits are on `VariantAnnotation`

File: src/main/resources/avro/bdg.avdl (hunk 19 above)
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: What's the representation we _want_ to have?

File: src/main/resources/avro/bdg.avdl (hunk 9 above)
Comment (devin-petersohn) [inline src/main/resources/avro/bdg.avdl]: I have tried to define the tables myself, and with the complexity involved in our nested objects, etc. it was a nightmare and I ended up putting it off for a while, especially for trying to just achieve compatibility with a single field. If we're expecting people to be able to use Impala with our schemas we need to keep this in mind.

File: src/main/resources/avro/bdg.avdl (hunk 13 above)
Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: As far as I know, likelihoods should be log10.

File: src/main/resources/avro/bdg.avdl (hunk 16 above)
Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: Agreed.  So we're talking about
```
likelihoods
nonReferenceLikelihoods
priorProbabilities  (is this useful?)
posteriorProbabilities
```
all log10-scaled?  Probabilities in VCF spec are "linear-scale probability in the
posterior distribution, which is Pr(Model$|$Data)".  Are you suggesting log10-scaling those too?

File: src/main/resources/avro/bdg.avdl (hunk 19 above)
Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: I don't have an opinion.

File: src/main/resources/avro/bdg.avdl (hunk 16 above)
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: Is there any driver to use log10 versus log?

File: src/main/resources/avro/bdg.avdl (hunk 22 above)
Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: As far as I know, likelihoods should be log10.

File: src/main/resources/avro/bdg.avdl (hunk 16 above)
Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: log10 is what I mean when I say "log-scaled", but then I'm a biologist, not a mathematician or statistician  :)

Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: Yeah, I mean, unless there's a significant reason to go `log10`, I would prefer natural log.

**Comment by jpdna** [comment]
Ping - what are the outstanding questions here?

**Comment by heuermh** [comment]
I'll push a commit tomorrow which addresses most of the review comments.  There are still some unresolved issues in the threads above (e.g. log10 vs. log).

File: src/main/resources/avro/bdg.avdl (hunk 16 above)
Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: Let me know if I have this right:

Field | Cardinality | Action 
----- | ---------- | ------  
`likelihoods` | GL Number=G Type=Float | Convert GL from log10 scale to log scale, optionally convert PL from phred scale to log scale
`nonReferenceLikelihoods` | Number="the ploidy at this site, plus 1" Type=Float | Log scaled, create new VCF key
`genotypePriors` | Number=G Type=Float | Log scaled, rename `priors`, create new VCF key
`genotypePosteriors` | Number=G Type=Float | Log scaled, rename `posteriors`, convert GP from log10 scale to log scale
`posteriorProbabilities` | GP Number=G Type=Float | Drop in favor of log scale `posteriors`

File: src/main/resources/avro/bdg.avdl [hunk 27]
```diff
@@ -748,166 +748,187 @@ record Variant {
...
-  union { null, float } readPositionRankSum = null;
+  //
+  // Not present in VCF specification yet present in Genotype
 
   /**
-   The log scale prior probabilities of the various genotype states at this site.
-   The number of elements in this array should be equal to the ploidy at this
-   site, plus 1.
+   Minimum number of reads seen at this site across samples when joint calling variants.
+   VCF genotype field key MIN_DP.
```
Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: This definition is not consistent with other definitions I've seen in the wild for `MIN_DP` (or `Min_DP`, or `min_DP`) VCF genotype field keys, e.g.

GATK
`##FORMAT=<ID=MIN_DP,Number=1,Type=Integer,Description="Minimum DP observed within the GVCF block">`

freebayes
`##FORMAT=<ID=MIN_DP,Number=1,Type=Integer,Description=\"Minimum depth in gVCF output block.\">`

File: src/main/resources/avro/bdg.avdl (hunk 16 above)
Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: This looks correct, with the exception that `nonReferenceLikelihoods` shouldn't go into a new VCF key, but that `nonReferenceLikelihoods` encodes the GL's for the symbolic non-ref (`*`) alt in a gVCF line.

Handling this gets a bit complex, but the way to think of it is that if `alternateAllele` is non-null, we get a VCF line with two alts (`alternateAllele`, and `*`). In the diploid case, we'd get GL's: `likelihoods(0), likelihoods(1), likelihoods(2), nonRefLikelihoods(1), nonRefLikelihoods(2), 0?`. I'm not entirely sure what should go in the last spot, which would be the `GT=1/2` likelihood, which I feel is not well defined. If `alternateAllele` is null, then we get a VCF line with just `*` as an alt and with GL=`nonRefLikelihoods`

File: src/main/resources/avro/bdg.avdl (hunk 27 above)
Comment (jpdna) [inline src/main/resources/avro/bdg.avdl]: I agree the comment documentation for this field should include "minimum depth within gVCF block" wording as do the examples @heuermh  lists above.

Comment (fnothaft) [inline src/main/resources/avro/bdg.avdl]: +1

Comment (heuermh) [inline src/main/resources/avro/bdg.avdl]: Is it that simple?

"[the minimum]...at this site across all samples" doesn't mean the same thing as "[the minimum]...across a set of non-variant sites grouped together into non-variant block records that represent intervals of sites for which the genotype quality (GQ) is within a certain range or band."  I don't believe we support
```
##GVCFBlock=minGQ=0(inclusive),maxGQ=5(exclusive)
##GVCFBlock=minGQ=20(inclusive),maxGQ=60(exclusive)
##GVCFBlock=minGQ=5(inclusive),maxGQ=20(exclusive)
##GVCFBlock=minGQ=60(inclusive),maxGQ=2147483647(exclusive)
```
header stuff like HaplotypeCaller outputs unless those are passed through unintentionally.

Or am I reading to much into the current description; it meant gVCF blocks all along?

**Comment by heuermh** [comment]
Fixes #106

**Comment by heuermh** [comment]
Most recent changes:
 * likelihoods and probabilities are log-scaled
 * new VCF genotype field keys `priors`, `nonReferenceLikelihoods`, and `fisherStrandBiasPValue`

I have reservations about log-scaling, in that users are not likely to look at the documentation, and will be confused when e.g. GL and `likelihoods` values do not match.

I'm still not sure what to do re: https://github.com/bigdatagenomics/bdg-formats/pull/108#discussion_r112486393 which may be hidden above.  We currently set the `variant` field and the flattened values when reading from VCF, so unless a user projects away the fields they don't want, we're using more RAM than necessary.

**Comment by heuermh** [comment]
Resolved thread around flatting variant fields on `Genotype` by removing `variant` field and adding `alternateAllele` and `referenceAllele`.

**Comment by heuermh** [comment]
Replaced by #176.
